dcicutils/obfuscation_utils.py

```diff
@@ -96,25 +154,30 @@ def obfuscate_dict(dictionary: dict, inplace: bool = False, show: bool = False,
     check_true(not obfuscated or is_obfuscated(obfuscated),
                message=f"If obfuscated= is supplied, it must be {OBFUSCATED_VALUE_DESCRIPTION}.")
 
-    def has_values_to_obfuscate(dictionary: dict) -> bool:
-        for key, value in dictionary.items():
-            if isinstance(value, dict):
-                if has_values_to_obfuscate(value):
-                    return True
-            elif isinstance(value, str) and should_obfuscate(key) and not is_obfuscated(value):
-                return True
-        return False
+    if show:
+        return item
+
+    orig_item = item
+
+    changed = StorageCell(False)
+
+    if not inplace:
+        item = copy.deepcopy(item)
+
+    def process_recursively(item):
+        # We only need to process non-atomic items recursively, since they are the only things
+        # that might conceivably be or contain a dictionary in need of obfuscation.
+        if isinstance(item, dict):
+            for key, value in item.items():
+                if should_obfuscate(key):
+                    changed.value = True
+                    item[key] = obfuscate(value, obfuscated=obfuscated)
+                else:
+                    process_recursively(value)
+        elif isinstance(item, list):
```

Comment:
I guess you didn't have a unit test for that. :)  Unit tests are responsible for noticing that you've changed something of consequence during a refactor. It's one of the primary reasons I'm so manic about them.

I'd want to think on this more. Do we know a use case/ It's semantically perilous to do an in-place change of the sort you mention because tuples are considered constant and you'd be changing the hash value of a constant. I suppose  since it's always newly made structure, no ill is going to come of that. I'll experiment with that if you think it's going to break something.  In practice, I'm not sure we ever store tuples in this way.

Reply:
My branch (obfuscation-utils-fix-20230330) had a couple test for tuples ... ```def test_obfuscate_dict_with_nested_tuple():

    d = {"abc": "123", "def": ({"ghi": "456", "jklsecret": "obfuscatethisvalue"}, 789), "jkl": "hello"}
    o = {"abc": "123", "def": ({"ghi": "456", "jklsecret": "<REDACTED>"}, 789), "jkl": "hello"}
    d_copy = copy.deepcopy(d)  # to make sure the original is not inadvertantly modified
    x = obfuscate_dict(d, obfuscated="<REDACTED>")
    assert x == o
    assert d == d_copy


def test_obfuscate_dict_with_tuple():

    d = ({"abc": "123", "def": ({"ghi": "456", "jklsecret": "obfuscatethisvalue"}, 789), "passwd": "hello"}, (1, 2, 3))
    o = ({"abc": "123", "def": ({"ghi": "456", "jklsecret": "<REDACTED>"}, 789), "passwd": "<REDACTED>"}, (1, 2, 3))
    d_copy = copy.deepcopy(d)  # to make sure the original is not inadvertantly modified
    x = obfuscate_dict(d, obfuscated="<REDACTED>")
    assert x == o
    assert d == d_copy```

Reply:
Tuples do come up the TRACE_ENABLED=true of submit-ontology (the auth tuple is output, secret in plain text).

Reply:
OK, I'm adding these tests and adjusting the type signature.